Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java

```diff
@@ -0,0 +1,138 @@
+        sendTelemetry(properties);
+    }
+
+    public static void afterExit(final IAzureOperation op) {
+        final long timestamp = System.currentTimeMillis();
+        final Map<String, String> properties = buildProperties(op);
+        properties.put(TELEMETRY_OP_TIMESTAMP, String.valueOf(timestamp));
+        properties.put(TELEMETRY_OP_ACTION, OP_ACTION_EXIT);
+        sendTelemetry(properties);
+    }
+
+    @NotNull
+    private static Map<String, String> buildProperties(final IAzureOperation op) {
+        final Deque<IAzureOperation> ctxOperations = AzureTaskContext.getContextOperations();
+        final Optional<IAzureOperation> parent = Optional.ofNullable(ctxOperations.peekFirst());
+        final Map<String, String> properties = new HashMap<>();
+        properties.put(TELEMETRY_OP_CONTEXT_ID, getCompositeId(ctxOperations, op));
+        properties.put(TELEMETRY_OP_ID, op.getId());
+        properties.put(TELEMETRY_OP_PARENT_ID, parent.map(IAzureOperation::getId).orElse("/"));
+        properties.put(TELEMETRY_OP_NAME, op.getName());
+        properties.put(TELEMETRY_OP_TYPE, op.getType());
+        return properties;
+    }
+
+    private static String getCompositeId(final Deque<? extends IAzureOperation> ops, IAzureOperation op) {
+        final List<IAzureOperation> revised = revise(ops);
+        revised.add(op);
+        return revised.stream().map(IAzureOperation::getId).collect(Collectors.joining("/", "/", ""));
+    }
+
+    /**
+     * get all ancestors until the last operation of ACTION type.
+     */
+    public static List<IAzureOperation> revise(Deque<? extends IAzureOperation> ops) {
+        final LinkedList<IAzureOperation> result = new LinkedList<>();
+        for (final IAzureOperation op : ops) {
+            result.addFirst(op);
+            if (op instanceof AzureOperationRef) {
+                final AzureOperation annotation = AzureOperationUtils.getAnnotation((AzureOperationRef) op);
+                if (annotation.type() == AzureOperation.Type.ACTION) {
+                    break;
+                }
+            }
+        }
+        return result;
+    }
+
+    @SneakyThrows
+    private static void sendTelemetry(final Map<String, String> properties) {
+        AppInsightsClient.create(PERFORMANCE, null, properties);
+    }
+
+    private static void writeToCsvFile(Map<String, String> properties) throws IOException {
+        final String val = String.format("%s, %s, %s, %s, %s, %s, %s",
+            properties.get(TELEMETRY_OP_CONTEXT_ID),
+            properties.get(TELEMETRY_OP_TIMESTAMP),
+            properties.get(TELEMETRY_OP_ID),
+            properties.get(TELEMETRY_OP_NAME),
+            properties.get(TELEMETRY_OP_TYPE),
+            properties.get(TELEMETRY_OP_ACTION),
+            properties.get(TELEMETRY_OP_PARENT_ID)
+        );
+        FileUtils.writeStringToFile(file, val + System.lineSeparator(), StandardCharsets.UTF_8, true);
```

Comment:
shall we use https://www.baeldung.com/apache-commons-csv instead?

Reply:
no, it's just for demo